packages/pick-version/src/lib.rs

```diff
@@ -0,0 +1,143 @@
+use std::collections::HashMap;
+
+use anyhow::{Context, Result};
+use chrono::prelude::*;
+use parse_package_arg::PackageArg;
+use semver::{Version, VersionReq};
+use ssri::Integrity;
+use thiserror::Error;
+
+pub struct Picker {
+    pub default_tag: String,
+}
+
+impl Default for Picker {
+    fn default() -> Self {
+        Picker {
+            default_tag: "latest".into(),
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum PickerError {
+    #[error("No matching version.")]
+    NoVersion,
+    #[error("Only Version, Tag, Range, and Alias package args are supported.")]
+    InvalidPackageArg,
+}
+
+// TODO - this type should probably live elsewhere.
+#[derive(Debug, Clone)]
+pub struct Packument {
+    pub created: DateTime<Utc>,
+    pub modified: DateTime<Utc>,
+    pub name: String,
+    pub require_tfa: bool,
+    pub tags: HashMap<String, Version>,
+    pub versions: HashMap<Version, Integrity>,
+}
+
+impl Default for Packument {
+    fn default() -> Self {
+        Packument {
+            created: Utc::now(),
+            modified: Utc::now(),
+            name: "foo".into(),
+            require_tfa: false,
+            tags: HashMap::new(),
+            versions: HashMap::new(),
+        }
+    }
+}
+
+impl Picker {
+    pub fn new() -> Self {
+        Picker::default()
+    }
+
+    pub fn default_tag(mut self, tag: String) -> Self {
+        self.default_tag = tag;
+        self
+    }
+
+    pub fn pick(&self, packument: &Packument, wanted: &PackageArg) -> Result<Integrity> {
+        if packument.versions.len() == 0 {
+            Err(PickerError::NoVersion).with_context(|| {
+                format!(
+                    "No versions were present in the packument for {}",
+                    packument.name
+                )
+            })?;
+        }
+        match wanted {
+            PackageArg::Alias { package, .. } => return self.pick(&packument, &package),
+            _ => (),
+        }
+
+        let mut target: Option<Version> = match wanted {
+            PackageArg::Version { version, .. } => Some(version.clone()),
+            PackageArg::Tag { tag, .. } => packument.tags.get(tag.as_str()).map(|v| v.clone()),
+            PackageArg::Range { .. } => None,
+            _ => Err(PickerError::InvalidPackageArg)
+                .with_context(|| format!("Received an unexpected PackageArg type: {:?}", wanted))?,
+        };
+
+        let tag_version = packument.tags.get(&self.default_tag).map(|v| v.clone());
```

Comment:
This brings to mind: do we want to preserve the magic "latest" tag behavior as-is? (That is, it is an invariant that packages must have a "latest" tag, and it _must_ point at the most recently published, non-deprecated version?) In the past the server had to move the `latest` tag around manually to represent "the most recently published package" – is this logic we could bake into the client? (Should we bake that logic into the client, so that the operation can be signed by the client?)

Reply:
(It's totally okay to defer this design question to a meeting & proceed with PR as-is – this line just got me thinking!)

Reply:
I think we should. Partly for compatibility reasons, partly because I don't hate it to begin with